engine/language-server/src/main/scala/org/enso/languageserver/text/CollaborativeBuffer.scala

```diff
@@ -79,60 +82,62 @@ class CollaborativeBuffer(
         path,
         client.clientId
       )
-      openBuffer(client, path)
+      openBuffer(client, path, Map.empty)
   }
 
   private def waitingForFileContent(
     rpcSession: JsonSession,
     replyTo: ActorRef,
-    timeoutCancellable: Cancellable
+    timeoutCancellable: Cancellable,
+    autoSave: Map[ClientId, Cancellable]
   ): Receive = {
     case ReadTextualFileResult(Right(content)) =>
-      handleFileContent(rpcSession, replyTo, content)
+      handleFileContent(rpcSession, replyTo, content, autoSave)
       unstashAll()
       timeoutCancellable.cancel()
 
     case ReadTextualFileResult(Left(failure)) =>
       replyTo ! OpenFileResponse(Left(failure))
       timeoutCancellable.cancel()
-      stop()
+      stop(autoSave)
 
     case IOTimeout =>
       replyTo ! OpenFileResponse(Left(OperationTimeout))
-      stop()
+      stop(autoSave)
 
     case _ => stash()
   }
 
   private def collaborativeEditing(
     buffer: Buffer,
     clients: Map[ClientId, JsonSession],
-    lockHolder: Option[JsonSession]
+    lockHolder: Option[JsonSession],
+    autoSave: Map[ClientId, Cancellable]
```

Comment:
Maybe I misunderstood the logic, but I wander why do we need to track the auto-save timeout per client? I thought we can just track that the buffer was not touched for `x` seconds by any of the clients, then save the contents to disc, and notify all the clients `clients: Map[ClientId, JsonSession]` about the performed action

Reply:
Ok, I read the implementation once again and noticed that we need to check that the client's version of the file matches the server version. Then tracking auto-save per client makes sense :+1:

Reply:
Yeah, that was the only motivation